### PR TITLE
Implement JSON r/o Web API for atomicapp

### PR DESCRIPTION
**Pull request by sub-mod**
PoC for https://github.com/projectatomic/atomicapp/issues/89
@vpavlin @goern 

daemon option:
atomicapp -d install aweiteka/guestbook-app

rest endpoints
curl -X GET http://localhost:5000/atomicapp-run/api/v1.0/status
curl -X POST http://localhost:5000/atomicapp-run/api/v1.0/quit

### CODE REVIEW

**Comment by vpavlin** [comment]
Can you please rebase on top of current master? There has been some cleanup done and this commit brings a lot of the mess back:)

**Comment by vpavlin** [comment]
Add something like 

```
atomicapp.config.globalCtx.addStatus(repr(ex))
```

At the line `109` (under `logger.error`) - to log error on failure

**Comment by vpavlin** [comment]
You have tabs and spaces mixed, please use spaces  (4 spaces = 1 tab) so that the code in project is consistent

File: atomicapp/utils.py [hunk 1]
```diff
@@ -20,6 +20,20 @@
...
     return str(val).lower() in true_values
 
+def addShelveKV(dbname, key, val):
+    with closing(shelve.open(dbname, writeback = True)) as s:
+        s[key] = val
+
+def appendShelveKV(dbname, key, val):
+    with closing(shelve.open(dbname, writeback = True)) as s:
+        temp = s[key]
+        temp.append(val)
```
Comment (vpavlin) [inline atomicapp/utils.py]: No need to use `temp`, simply call

```
s[key].append(val)
```

File: atomicapp/cli/main.py [hunk 2]
```diff
@@ -106,8 +109,20 @@ def run(self):
...
+    start_restapp(debug=False, port=5000)
+
+
+def start_cli():
     cli = CLI()
     cli.run()
 
+def start_restapp(debug, port):
+    rest = atomicapp.rest.Rest()
+    rest.addResource(atomicapp.rest.StatusList, '/atomicapp-run/api/v1.0/status')
```
Comment (vpavlin) [inline atomicapp/cli/main.py]: This should be probably part `rest.py`

File: atomicapp/utils.py (hunk 1 above)
Comment (sub-mod) [inline atomicapp/utils.py]: in shelve you have to do this. you need to fetch the entire value and  append and save the entire object back again.

Comment (vpavlin) [inline atomicapp/utils.py]: interesting..I tried that and it worked..

Comment (sub-mod) [inline atomicapp/utils.py]: ah ok i have  writeback = True ,so you dont need it. writeback = True ; will cache the data and rewrite it.

**Comment by vpavlin** [comment]
Can you please just commit changes rather than doing a rebase everytime? This is really hard to review - I can't see the differences between your commit if you do rebase..

**Comment by sub-mod** [comment]
ok , sure.

**Comment by goern** [comment]
@sub-mod could you add `-p 5000:5000` to `atomicapp.sh` so that it supports your REST server. And change the API version to `__ATOMICAPPVERSION__` it is defined in https://github.com/projectatomic/atomicapp/blob/master/atomicapp/constants.py

**Comment by goern** [comment]
please also put at least `id` from the nulecule file in each response, so that clients know which atomicapp instance is responding, maybe the `metadata` object too

**Comment by goern** [comment]
beside of that comments it looks good! please what out for landscape comments and keep up the good work! thx.

**Comment by vpavlin** [comment]
Please file namespace commit as separate PR

**Comment by sub-mod** [comment]
@vpavlin  https://github.com/projectatomic/atomicapp/pull/123

**Comment by vpavlin** [comment]
This patch: https://vpavlin.fedorapeople.org/0001-Fix-rest-API-feature.patch should fix all issues and make this PR acceptable, please apply it  and push to you branch

```
curl -O https://vpavlin.fedorapeople.org/0001-Fix-rest-API-feature.patch
git am 0001-Fix-rest-API-feature.patch
```

**Comment by sub-mod** [comment]
@goern kindly review

**Comment by vpavlin** [comment]
LGTM 

Note: Neither I, nor @sub-mod think this is a final perfect solution, but it's good enough to have a working example of integration with Cockpit

**Comment by sub-mod** [comment]
+1 @vpavlin 's comment
the way we do addStatus is a duplication of logging efforts. IMHO i think we should a refactor of the code flow , add the some method descriptors to avoid duplication.But It's ok now, I think. We need to  figure the way out later.

**Comment by goern** [comment]
LOKTM, please test accessing the REST API while `atomic install projectatomic/wordpress-centos7-atomicapp` (or zookeeper or whatever atomicapp) is running, please update the atomicapp base image. and I vote for keeping this in :dev up until 0.5.0 aka https://github.com/projectatomic/atomicapp/milestones/Red%20Hat%20Summit%202015

**Comment by sub-mod** [comment]
@goern is port 5000 value acceptable ?

**Comment by stefwalter** [comment]
@goern @sub-mod I'm not sure I understand how this rest API is meant to be located. Will it always listen on port 5000 and conflict with any service listening on that port on the host or other container on that port? Is it by design that only one such atomicapp installer can be active at a time?

**Comment by vpavlin** [comment]
@stefwalter Good question. I don't think we should force a single instance of atomicapp at a time.

@sub-mod Is it possible to remove port binding to the host port and use <container_ip>:<port> instead of localhost:<port> for you in your cockpit work?

**Comment by goern** [comment]
@vpavlin @sub-mod as we use `--net=host` for atomicapp ( #165 ) container, I think the restserver cant listen on any container_ip

**Comment by sub-mod** [comment]
since atomicapp will be invoked within the atomicapp container via atomic command there will be no port conflict because each instance of the atomic can have a different exposed port number.

**Comment by stefwalter** [comment]
> since atomicapp will be invoked within the atomicapp container via atomic command there will be no port conflict because each instance of the atomic can have a different exposed port number.

How would you know which container corresponds to the command you just launched?

>  curl -X POST http://localhost:5000/atomicapp-run/api/v1.0/quit

This is unauthenticated. Anyone on the machine can quit any active atomicapp installer. As I've mentioned before, I think the REST API is needlessly complex way to get status information. It's possible to make it work, but it's going to be a lot of pain finishing it up.

**Pull request closed without merging**